events/tag/src/module/attack.rs

```diff
@@ -150,368 +185,128 @@ impl Module for AttackModule {
-                                        // player died, increment kill count
-                                        kill_count.kill_count += 1;
-
-                                        target.set::<Team>(*origin_team);
-
-                                        origin_xp.amount = (f32::from(target_xp.amount)*0.5) as u16;
-                                        target_xp.amount = (f32::from(target_xp.amount)/3.) as u16;
-
-                                        return;
-                                    }
-
-                                    // Calculate velocity change based on attack direction
-                                    let this = **target_position;
-                                    let other = **origin_pos;
-
-                                    let dir = (this - other).normalize();
-
-                                    let knockback_xz = 8.0;
-                                    let knockback_y = 6.432;
-
-                                    let new_vel = Vec3::new(
-                                        dir.x * knockback_xz / 20.0,
-                                        knockback_y / 20.0,
-                                        dir.z * knockback_xz / 20.0
-                                    );
-
-                                    target_velocity.0 += new_vel;
-
-                                    // https://github.com/valence-rs/valence/blob/8f3f84d557dacddd7faddb2ad724185ecee2e482/examples/ctf.rs#L987-L989
-                                },
-                            );
+                    for event::AttackEntity { origin, target, damage } in event_queue.drain() {
+                        let origin = world.entity_from_id(origin);
+                        let target = world.entity_from_id(target);
+
+                        if check_and_update_immunity(current_tick, &target) {
+                            // no damage; the target is immune
+                            continue;
+                        }
+
+                        let origin_team = origin.get::<&Team>(|team| *team);
+                        let target_team = target.get::<&Team>(|team| *team);
+
+                        let origin_pos = origin.get::<&Position>(|position| *position);
+                        let target_pos = target.get::<&Position>(|position| *position);
+
+                        let origin_connection = origin.get::<&ConnectionId>(|connection_id| *connection_id);
+                        let target_connection = target.get::<&ConnectionId>(|connection_id| *connection_id);
+
+
+                        if origin_team == target_team {
+                            let msg = "§cCannot attack teammates";
+                            let pkt_msg = play::GameMessageS2c {
+                                chat: msg.into_cow_text(),
+                                overlay: false,
+                            };
+
+                            compose.unicast(&pkt_msg, origin_connection, system).unwrap();
```

Comment:
Skip the attack event if it's a teammate:
```suggestion
                            compose.unicast(&pkt_msg, origin_connection, system).unwrap();
                            continue;
```